### PR TITLE
FEATURE: Starterfile Parser

### LINKED ISSUE #3: Starterfile definition
Define the structures and function of a Starterfile

### LINKED ISSUE #5: StarterFile Parsing Tools
### Description:
---
Related to issue #3, this issue is specifically for the functionality used when parsing through a StarterFile. This is the essential part of creating the path

### Tasklist:
---
- [x] Parse tools
- [x] Parse Modules
- [x] Parse & Initialize Env Variables
 _add more if needed_

### LINKED ISSUE #6: StarterFile Documentation
### Description
---
Create documentation around the StarterFile that will assist other developers who will work with this system. Logic as to why you approached things the way you did would be useful to include as well but is up to you.

_Feel free to adjust this as needed..._

### PR DESCRIPTION
**Pull request by trentonyo**
Starterfile Parser
===========

> **ASSOCIATED ISSUES:**
>
> 
> 
> 
>
> 
> 
>  
> - Closes #3
> - Closes #5
> - Closes #6 

Description
-----------

🤖📂 **Starterfile:** Defined a structure and parser for Starterfiles. These files define the tools and modules used in a path and, in some situations, may define the entirety of the framework. This feature is an important aspect of Template-as-Code and will allow for easy replication and reuse of important tools and frameworks.

This feature is yet to be integrated into the CLI, but in order to test it one can run `starterfile.py` and play around with the main function. Unit tests for the parser and integration tests which cover several defined Starterfiles are included.

Definition of Done
------------------

* [x] Changes are approved
* [x] Contains no breaking changes
* [ ] Changes are implemented in all components
  * Integration with the CLI will be a separate PR.
* [x] Each component associated with issue is updated (no stubs)
* [x] Each new feature has unit tests written and confirmed
* [x] All tests pass, no regressions
* [x] All documentation is updated, including release notes

### CODE REVIEW

File: startout/starterfile.py [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+            # Destroy modules which were successfully installed if specified
+            if teardown_on_failure:
+                print("Rolling back successfully installed modules:",
+                      successful_modules,
+                      file=sys.stderr)
+
+                destroyed_modules = []
+                for module in [module for module in self.modules if module.name in successful_modules]:
+                    if not module.destroy():
+                        # TODO handle failure to destroy better
```
Comment (Gehrkej) [inline startout/starterfile.py:245]: Found this TODO, unsure if we want to merge it down or handle it first

**Review by Gehrkej — COMMENTED** [review]
Only one comment, I don't have a lot of experience with these libraries so a lot of is foreign to me but I am confident in your abilities and if it works for you and has your stamp of approval than you are good to merge 😄

File: startout/starterfile.py (hunk 1 above)
Comment (trentonyo) [reply startout/starterfile.py:245]: I think that later on down the road we can define authors of modules/help docs that could be useful in these scenarios, but if a module/tool fails to destroy then it's likely outside of the scope of the CLI to resolve. Thank you for finding this! Once we integrate todoon it'll generate an issue for us lol

**Review by Gehrkej — APPROVED** [review]
Ready to Merge!

**Pull request merged**